### PR TITLE
#20 Reimplement create/update

### LINKED ISSUE #20: Reimplement create/update
For now, we are returning the whole object from a repository after `create` action. We should return only the newly created identifier. Frontend application will get the newly created object then.

### PR DESCRIPTION
**Pull request by devmonte**
resolve #20

### CODE REVIEW

**Comment by devmonte** [comment]
resolve #20

File: 1 Presentation layer/CandidateEvaluator.Client/Pages/Category/Category.cshtml [hunk 1]
```diff
@@ -32,7 +32,7 @@ protected override async Task OnInitAsync()
 
 private async Task Save()
 {
-    await Http.AuthorizedPostJsonAsync<Contract.Models.Category>($"api/Category/{Id}", _categoryModel);
+    await Http.AuthorizedPostJsonAsync<Guid>($"api/Category/{Id}", _categoryModel);
```
Comment (meanin) [inline 1 Presentation layer/CandidateEvaluator.Client/Pages/Category/Category.cshtml]: I believe that to be clear with `Get on Post` rule, you should get the id from post output and then redirect/navigate to the new path. Like in a CreateCategory page.

File: 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace CandidateEvaluator.Core.Dispatchers
+{
+    public class CommandDispatcher : ICommandDispatcher
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        public CommandDispatcher(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs]: This is an antipattern called service locator. Is there a way to get it rid off?

File: 2 Business layer/CandidateEvaluator.Core/Dispatchers/QueryDispatcher.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace CandidateEvaluator.Core.Dispatchers
+{
+    public class QueryDispatcher : IQueryDispatcher
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        public QueryDispatcher(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Dispatchers/QueryDispatcher.cs]: Again, service locator.

File: 3 Domain layer/CandidatesEvaluator.Contract/Handlers/ICommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using CandidateEvaluator.Contract.Commands;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CandidateEvaluator.Contract.Handlers
+{
+    public interface ICommandHandler<T> where T : ICommand
```
Comment (meanin) [inline 3 Domain layer/CandidatesEvaluator.Contract/Handlers/ICommandHandler.cs]: Make it [contravariant](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/creating-variant-generic-interfaces) with `in` operator like below:
`public interface ICommandHandler<in T> where T : ICommand`

File: 3 Domain layer/CandidatesEvaluator.Contract/Handlers/IQueryHandler.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿using CandidateEvaluator.Contract.Queries;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CandidateEvaluator.Contract.Handlers
+{
+    public interface IQueryHandler<TQuery, TResult> where TQuery : IQuery<TResult>
```
Comment (meanin) [inline 3 Domain layer/CandidatesEvaluator.Contract/Handlers/IQueryHandler.cs]: Make it both [variant and contravariant](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/creating-variant-generic-interfaces) with in/out operators like below:
`public interface IQueryHandler<in TQuery, out TResult> where TQuery : IQuery<TResult>`

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Commands/CreateCategoryHandler.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Commands
+{
+    public class CreateCategoryHandler : ICommandHandler<CreateCategory>
+    {
+        private readonly ICategoryService _categoryService;
+
+        public CreateCategoryHandler(ICategoryService categoryService)
+        {
+            _categoryService = categoryService;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Commands/CreateCategoryHandler.cs]: I believe that handlers should operate straight on repositories/context. The service layer is redundant here.

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Commands/DeleteCategoryHandler.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Commands
+{
+    public class DeleteCategoryHandler : ICommandHandler<DeleteCategory>
+    {
+        private readonly ICategoryService _categoryService;
+
+        public DeleteCategoryHandler(ICategoryService categoryService)
+        {
+            _categoryService = categoryService;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Commands/DeleteCategoryHandler.cs]: I believe that handlers should operate straight on repositories/context. The service layer is redundant here.

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Commands/UpdateCategoryHandler.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Commands
+{
+    public class UpdateCategoryHandler : ICommandHandler<UpdateCategory>
+    {
+        private readonly ICategoryService _categoryService;
+
+        public UpdateCategoryHandler(ICategoryService categoryService)
+        {
+            _categoryService = categoryService;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Commands/UpdateCategoryHandler.cs]: I believe that handlers should operate straight on repositories/context. The service layer is redundant here.

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetAllCategoriesHandler.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Queries
+{
+    public class GetAllCategoriesHandler : IQueryHandler<GetAllCategories, List<Category>>
+    {
+        private readonly ICategoryService _categoryService;
+
+        public GetAllCategoriesHandler(ICategoryService categoryService)
+        {
+            _categoryService = categoryService;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetAllCategoriesHandler.cs]: I believe that handlers should operate straight on repositories/context. The service layer is redundant here.

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetAllUserActivitiesHandler.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Queries
+{
+    public class GetAllUserActivitiesHandler : IQueryHandler<GetAllUserActivities, List<RecentActivity>>
+    {
+        private readonly IUserRecentActivityService _activityService;
+
+        public GetAllUserActivitiesHandler(IUserRecentActivityService activityService)
+        {
+            _activityService = activityService;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetAllUserActivitiesHandler.cs]: I believe that handlers should operate straight on repositories/context. The service layer is redundant here.

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetCategoryHandler.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Queries
+{
+    public class GetCategoryHandler : IQueryHandler<GetCategory, Category>
+    {
+        private readonly ICategoryService _categoryService;
+
+        public GetCategoryHandler(ICategoryService categoryService)
+        {
+            _categoryService = categoryService;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetCategoryHandler.cs]: I believe that handlers should operate straight on repositories/context. The service layer is redundant here.

File: 3 Domain layer/CandidatesEvaluator.Contract/Queries/IQuery.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CandidateEvaluator.Contract.Queries
+{
+    public interface IQuery<in TResult>
```
Comment (meanin) [inline 3 Domain layer/CandidatesEvaluator.Contract/Queries/IQuery.cs]: nice!

File: 3 Domain layer/CandidatesEvaluator.Contract/Queries/QueryBase.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CandidateEvaluator.Contract.Queries
+{
+    public class QueryBase
```
Comment (meanin) [inline 3 Domain layer/CandidatesEvaluator.Contract/Queries/QueryBase.cs]: This one class is redundant. OwnerId prop could be placed in all queries. There is no gain to extract it as a base class.

**Comment by meanin** [comment]
Furthermore, update readme, please.

File: 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs (hunk 2 above)
Comment (devmonte) [reply 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs]: I was aware of that, but dispatcher is really small. Another way could be to inject a list of all command handlers and then choose one, but we generic interface so default DI container doesn't allow for that. I'll research if other DI containers have that feature.

Comment (meanin) [reply 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs]: Factory?

Comment (meanin) [reply 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs]: If you want to inject a list of handlers this is possible whitout using different IoC

Comment (devmonte) [reply 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs]: Default Asp.Net Core IoC is able to inject a list of generics?

**Comment by devmonte** [comment]
Based on examples form the internet: [DevMentors](https://github.com/devmentors/DNC-DShop.Services.Products/blob/master/src/DShop.Services.Products/Handlers/CreateProductHandler.cs), this [blog post](https://exceptionnotfound.net/implementing-cqrs-in-net-part-2-handling-commands-and-events/) and [this](https://buildplease.com/pages/fpc-10/) I decided to move logic from Services to Handlers.

File: 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs (hunk 2 above)
Comment (devmonte) [reply 2 Business layer/CandidateEvaluator.Core/Dispatchers/CommandDispatcher.cs]: [Mentors](https://github.com/devmentors/DNC-DShop.Common/blob/master/src/DShop.Common/Dispatchers/QueryDispatcher.cs) handle that problem in the same way.

File: 3 Domain layer/CandidatesEvaluator.Contract/Handlers/IQueryHandler.cs (hunk 5 above)
Comment (devmonte) [reply 3 Domain layer/CandidatesEvaluator.Contract/Handlers/IQueryHandler.cs]: image
Nope :(

Comment (meanin) [reply 3 Domain layer/CandidatesEvaluator.Contract/Handlers/IQueryHandler.cs]: `IQuery<out T>` ?

File: 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetAllCategoriesHandler.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace CandidateEvaluator.Core.Handlers.Queries
+{
+    public class GetAllCategoriesHandler : IQueryHandler<GetAllCategories, List<Category>>
+    {
+        private readonly ICategoryRepository _modelRepository;
+
+        public GetAllCategoriesHandler(ICategoryRepository modelRepository)
+        {
+            this._modelRepository = modelRepository;
```
Comment (meanin) [inline 2 Business layer/CandidateEvaluator.Core/Handlers/Queries/GetAllCategoriesHandler.cs]: Where did `this` come from?

**Pull request merged**